File: src/brace_sync/crypto.py
"""Client-side encryption for vault objects.

The vault encrypts every payload before it reaches Cloudflare R2, GitHub, or
any other storage backend. This implementation avoids mandatory Python package
crypto dependencies by using Python's standard-library scrypt/HMAC primitives
and the system OpenSSL command for AES-256-CBC encryption.

The envelope uses encrypt-then-MAC: ciphertext and authenticated associated
data are verified with HMAC-SHA256 before any decryption is attempted.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import shutil
import subprocess
from dataclasses import dataclass
from typing import Any

from .errors import EncryptionError

ENVELOPE_SCHEMA_VERSION = 1
ALGORITHM = "aes-256-cbc-hmac-sha256"
KDF = "scrypt"
SALT_BYTES = 16
IV_BYTES = 16
KEY_BYTES = 64
SCRYPT_N = 2**14
SCRYPT_R = 8
SCRYPT_P = 1
OPENSSL = "openssl"


def _b64encode(value: bytes) -> str:
    return base64.b64encode(value).decode("ascii")


def _b64decode(value: str) -> bytes:
    return base64.b64decode(value.encode("ascii"), validate=True)


def canonical_json(value: Any) -> bytes:
    """Serialize JSON deterministically for authenticated associated data."""
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode("utf-8")


def derive_key(passphrase: str, salt: bytes) -> bytes:
    """Derive encryption and authentication keys from a passphrase and salt."""
    if not passphrase:
        raise EncryptionError("A non-empty vault passphrase is required")
    return hashlib.scrypt(
        passphrase.encode("utf-8"),
        salt=salt,
        n=SCRYPT_N,
        r=SCRYPT_R,
        p=SCRYPT_P,
        dklen=KEY_BYTES,
    )


def _require_openssl() -> str:
    executable = shutil.which(OPENSSL)
    if not executable:
        raise EncryptionError("OpenSSL is required for vault encryption but was not found on PATH")
    return executable


def _openssl_crypt(payload: bytes, key: bytes, iv: bytes, decrypt: bool = False) -> bytes:
    command = [
        _require_openssl(),
        "enc",
        "-aes-256-cbc",
        "-K",
        key.hex(),
        "-iv",
        iv.hex(),
    ]
    if decrypt:
        command.append("-d")
    result = subprocess.run(command, input=payload, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=False)
    if result.returncode != 0:
        raise EncryptionError("OpenSSL encryption operation failed")
    return result.stdout


def _mac_payload(aad: dict[str, Any], iv: bytes, ciphertext: bytes) -> bytes:
    return canonical_json(aad) + b"\n" + iv + b"\n" + ciphertext


@dataclass(frozen=True)
class EncryptedEnvelope:
    """JSON-serializable encrypted object envelope."""

    algorithm: str
    kdf: str
    salt: str
    iv: str
    ciphertext: str
    tag: str
    aad: dict[str, Any]
    schema_version: int = ENVELOPE_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "algorithm": self.algorithm,
            "kdf": self.kdf,
            "salt": self.salt,
            "iv": self.iv,
            "ciphertext": self.ciphertext,
            "tag": self.tag,
            "aad": self.aad,
        }

    def to_json_bytes(self) -> bytes:
        return json.dumps(self.to_dict(), sort_keys=True, indent=2).encode("utf-8") + b"\n"

    @classmethod
    def from_json_bytes(cls, payload: bytes) -> "EncryptedEnvelope":
        try:
            value = json.loads(payload.decode("utf-8"))
            return cls(
                schema_version=int(value["schema_version"]),
                algorithm=str(value["algorithm"]),
                kdf=str(value["kdf"]),
                salt=str(value["salt"]),
                iv=str(value["iv"]),
                ciphertext=str(value["ciphertext"]),
                tag=str(value["tag"]),
                aad=dict(value.get("aad", {})),
            )
        except (KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
            raise EncryptionError("Invalid encrypted envelope") from exc


def encrypt_payload(plaintext: bytes, passphrase: str, aad: dict[str, Any]) -> bytes:
    """Encrypt bytes into a JSON envelope."""
    salt = os.urandom(SALT_BYTES)
    iv = os.urandom(IV_BYTES)
    key_material = derive_key(passphrase, salt)
    encryption_key = key_material[:32]
    mac_key = key_material[32:]
    ciphertext = _openssl_crypt(plaintext, encryption_key, iv)
    tag = hmac.new(mac_key, _mac_payload(aad, iv, ciphertext), hashlib.sha256).digest()
    envelope = EncryptedEnvelope(
        algorithm=ALGORITHM,
        kdf=KDF,
        salt=_b64encode(salt),
        iv=_b64encode(iv),
        ciphertext=_b64encode(ciphertext),
        tag=_b64encode(tag),
        aad=aad,
    )
    return envelope.to_json_bytes()


def decrypt_payload(envelope_bytes: bytes, passphrase: str, expected_aad: dict[str, Any] | None = None) -> bytes:
    """Decrypt a JSON envelope and optionally require exact associated data."""
    envelope = EncryptedEnvelope.from_json_bytes(envelope_bytes)
    if envelope.schema_version != ENVELOPE_SCHEMA_VERSION:
        raise EncryptionError(f"Unsupported envelope schema version: {envelope.schema_version}")
    if envelope.algorithm != ALGORITHM:
        raise EncryptionError(f"Unsupported encryption algorithm: {envelope.algorithm}")
    if envelope.kdf != KDF:
        raise EncryptionError(f"Unsupported key derivation function: {envelope.kdf}")
    if expected_aad is not None and envelope.aad != expected_aad:
        raise EncryptionError("Encrypted envelope associated data did not match expected metadata")

    try:
        salt = _b64decode(envelope.salt)
        iv = _b64decode(envelope.iv)
        ciphertext = _b64decode(envelope.ciphertext)
        supplied_tag = _b64decode(envelope.tag)
    except ValueError as exc:
        raise EncryptionError("Invalid base64 in encrypted envelope") from exc

    key_material = derive_key(passphrase, salt)
    encryption_key = key_material[:32]
    mac_key = key_material[32:]
    expected_tag = hmac.new(mac_key, _mac_payload(envelope.aad, iv, ciphertext), hashlib.sha256).digest()
    if not hmac.compare_digest(supplied_tag, expected_tag):
        raise EncryptionError("Unable to authenticate payload with the supplied passphrase")
    return _openssl_crypt(ciphertext, encryption_key, iv, decrypt=True)

File: shared/src/crypto.js
const ENCODER = new TextEncoder();
const DECODER = new TextDecoder();

export const ENVELOPE_VERSION = 1;
export const DEFAULT_PBKDF2_ITERATIONS = 100_000;
export const AES_GCM_IV_BYTES = 12;
export const SALT_BYTES = 16;

function subtleCrypto() {
  if (!globalThis.crypto?.subtle) {
    throw new Error('Web Crypto API is required for client-side encryption');
  }
  return globalThis.crypto.subtle;
}

export function bytesToBase64(bytes) {
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(bytes).toString('base64');
  }
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

export function base64ToBytes(value) {
  if (typeof Buffer !== 'undefined') {
    return new Uint8Array(Buffer.from(value, 'base64'));
  }
  const binary = atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes;
}

export function randomBytes(length) {
  const bytes = new Uint8Array(length);
  globalThis.crypto.getRandomValues(bytes);
  return bytes;
}

export function randomId(prefix = '') {
  const id = bytesToBase64(randomBytes(16)).replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '');
  return `${prefix}${id}`;
}

async function importPassword(password) {
  return subtleCrypto().importKey('raw', ENCODER.encode(password), 'PBKDF2', false, ['deriveKey']);
}

export async function deriveAesKey(password, salt, iterations = DEFAULT_PBKDF2_ITERATIONS) {
  if (!password) throw new Error('A non-empty password is required');
  const baseKey = await importPassword(password);
  return subtleCrypto().deriveKey(
    { name: 'PBKDF2', salt, iterations, hash: 'SHA-256' },
    baseKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt']
  );
}

export async function encryptJson(plaintextValue, password, aad = {}) {
  const salt = randomBytes(SALT_BYTES);
  const iv = randomBytes(AES_GCM_IV_BYTES);
  const key = await deriveAesKey(password, salt);
  const plaintext = ENCODER.encode(JSON.stringify(plaintextValue));
  const additionalData = ENCODER.encode(JSON.stringify(aad));
  const ciphertext = await subtleCrypto().encrypt({ name: 'AES-GCM', iv, additionalData }, key, plaintext);

  return {
    version: ENVELOPE_VERSION,
    algorithm: 'AES-GCM',
    kdf: 'PBKDF2-SHA256',
    iterations: DEFAULT_PBKDF2_ITERATIONS,
    salt: bytesToBase64(salt),
    iv: bytesToBase64(iv),
    aad,
    ciphertext: bytesToBase64(new Uint8Array(ciphertext))
  };
}

export async function decryptJson(envelope, password, expectedAad = undefined) {
  if (envelope.version !== ENVELOPE_VERSION) {
    throw new Error(`Unsupported envelope version: ${envelope.version}`);
  }
  if (envelope.algorithm !== 'AES-GCM') {
    throw new Error(`Unsupported algorithm: ${envelope.algorithm}`);
  }
  if (expectedAad && JSON.stringify(envelope.aad) !== JSON.stringify(expectedAad)) {
    throw new Error('Encrypted envelope associated data did not match');
  }

  const salt = base64ToBytes(envelope.salt);
  const iv = base64ToBytes(envelope.iv);
  const ciphertext = base64ToBytes(envelope.ciphertext);
  const key = await deriveAesKey(password, salt, envelope.iterations);
  const additionalData = ENCODER.encode(JSON.stringify(envelope.aad ?? {}));
  const plaintext = await subtleCrypto().decrypt({ name: 'AES-GCM', iv, additionalData }, key, ciphertext);
  return JSON.parse(DECODER.decode(plaintext));
}
